### ISSUE TITLE
Artisan server/client

### ISSUE THREAD

**Issue by rpaulo** [artisan-roaster-scope/artisan#169]
It would be interesting if we could separate the backend from the frontend.  This would allow us to create an artisan server which could run on an RPI connected directly to the roaster and then run the frontend on a laptop.  They could communicate over WiFi or Bluetooth.

**Comment by MAKOMO** [comment]
There are users running Artisan on their RPi operating it via an iPad connected via VNC to that RPi. Raspian now comes out of the box with an VNC server. One level deeper, Matplotlib has a Web backend thus I have no experience with that and do not know if it supports enough features for running Artisan that way. I don't know if Qt has a Web-based backend.

**Comment by rpaulo** [comment]
I see a couple of drawbacks with that approach.

* The RPI is not a fast computer so I suspect using Artisan on it is a bit slow.  Artisan + VNC will probably make it even slower
* With a server/client architecture, we could run the Artisan server on even smaller computers.  Imagine we define a protocol for Artisan to communicate with the backend.  It could then implemented on an Arduino with an Ethernet shield.
* When we roast, we want to save the profiles somewhere.  If Artisan is running on the RPI, we have to save the profiles to the RPI first and the maybe copy them to a laptop.

**Comment by MAKOMO** [comment]
You are right, that a cleaner architecture is to use custom protocol, thus this involves a lot more development efforts.

To my experience the RPi3 is fast enough for everything around Artisan, including a VNC over a local WiFi.

Still, a small backend just handling the communication with external devices and abstracting away the device specifics could lead to a modular architecture the whole project could benefit from as it would separate concerns.

Should we tag this 2.0?

**Comment by rpaulo** [comment]
The method I was thinking of how this should work would be:

1) Artisan client is setup to connect to the server
2) After connecting, server sends the capabilities of the roaster (drum control, heater control, fan control) and LCD config (need to think about this a bit more). Alarms/events can be configured on both sides.
3) Artisan client configures the UI and asks the server to start streaming data

We can implement the communication channel using JSON since it's something that can be readily used by small computers like the Arduino.

**Comment by MAKOMO** [comment]
Interesting, but how are the capabilities on the server configured. By a config file modified by the user?
Yes, JSON via WebSockets might be the way to go (assuming the Arduino has a network connection). Still this sounds like a bigger project. 

Note that Phidgets and Yoctopuce modules already support a remote access as well as the Arduino TC4 by just adding a $1 Bluetooth module to transfer its serial stream via BT. There is also an ESP32-based hardware module for Artisan with 2 TC input channels and 2 motor control channels that talks MODBUS TCP via its Wifi connection. 

I have to think more about this all...

**Comment by PDWest** [comment]
Although it's not exactly what's asked for in the OP, but I have a simple modbus server that runs on a headless RPi zero.  It's got a couple of max31856 digital thermocouple inputs wired to the SPI bus.  For heater control I've got a Nodelynk I2C expansion bus with a 'slow' PWM for heater control plus a yet to be integrated with Artisan ambient temp/humidity sensor on it.  This little RPi connects to my roaster, and Artisan connects to it via WiFi using the standard modbus interface.   No refactoring of Artisan required.

**Comment by doza62** [comment]
Hi, has any work been done on this? If someone can point me to the right direction I could start building a simple api or cli utility for basic commands.

First, I believe configurations should be the first focus. See and set configurations through cli or simple web interface can help get the ball rolling towards a headless server and custom interfaces.

**Comment by PDWest** [comment]
I'm still using my RPi headless modbus server and it's working fine.  For my application, Artisan already has a full modbus interface, so it's not clear why an additional cli or web interface is needed or desired.

**Comment by doza62** [comment]
I'm not very familiar with modbus. How is your setup? Artisan on RPi and you connect with a custom UI through modbus to check on it and configure things?

**Comment by PDWest** [comment]
I use a little rpi zero-w at the roaster as a wifi-enabled Modbus TCP server.  My code is just a few lines of python that leverages an existing modbus package and a few other packages for data i/o (to the roaster).  When I start my roaster, the pi boots, connects to my home wifi and then waits for Artisan to connect through wifi.  There is no 'custom UI', I just use the excellent and extensive Artisan modbus interface features.  I'm currently running Artisan on a linux laptop, but before this, I had used a Macbook for years and have also used Windows.  It's very nice to have the wifi connection instead of having the Artisan machine tethered to the roaster with some kind of hard-wired connection (ethernet, usb).
